lib/quicly.c

```diff
@@ -340,6 +387,23 @@ struct st_quicly_conn_t {
          */
         uint8_t should_rearm_on_send : 1;
     } idle_timeout;
+    /**
+     * connection IDs we issue to the peer
+     */
+    struct {
+        /**
+         * list of struct st_quicly_pending_new_cid_t
+         */
+        struct st_quicly_issued_cid_t cids[QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT];
+        /**
+         * # of elements in cids whose state is PENDING
+         */
+        uint64_t num_pending;
+        /**
+         * # of retired CIDs
+         */
+        uint64_t num_retired;
+    } issued_cid;
```

Comment:
Here's what I understood from your comments (beside using `pending_flows`):
Instead of trying to find an empty (i.e. `IDLE`) slot each time, introduce a simple principle that elements `[0, ..., N-1]` shall always be non-idle. Any `IDLE` entry between `[0, N-1]` shall be converted to `PENDING` so that corresponding NCID frame will be sent in the next send round.

If I understood correctly, I think that's worth pursuing. Let me try to implement it.

Reply:
Implemented the new logic based on my understanding above.

Reply:
One thing maybe worth discussing is the way I handle the bit field. Currently I use one bit to represent if there is any pending NCID or RCID. I did not introduce separate bits for NCID and RCID, because it would need to extend `pending_flows` to `uint16_t`. Also, I thought NCID/RCID were rarely sent, so it would make sense to consolidate then into one bit.

Let me know what you think.